### DISCUSSION TITLE
Use of validation data for early stopping and learning curve

### DISCUSSION THREAD

**Discussion by ecetsn** [alan-turing-institute/sktime#7014] [Q&A]
Hello,

I'm looking for guidance on monitoring validation loss when using the SKTime CNN for class methods. I am currently using CNN for multivariate classification. Specifically, I’m trying to find a way to fit the model and track the loss on validation data throughout the epochs. I know I can use Keras EarlyStopping, but it only monitors training loss. Is there a method to obtain the learning curve for validation loss or to prevent the model from overfitting to the training data? The current fit method does not take validation data as a parameter.

Best,
Ece

**Comment by BlankAdventure** [comment]
Hi, I'm not sure if you're looking for an answer in here in terms of sktime or keras. In Keras, you just need to do something like:

```python
early_stopper = EarlyStopping(monitor='val_loss', mode='min', verbose=True, patience=8, restore_best_weights=True)
reduce_lr_callback = ReduceLROnPlateau(monitor='val_loss', factor=0.5, min_lr=1e-8, patience=5, verbose=True)
callbacks = [reduce_lr_callback, early_stopper]
history = autoencoder.fit(callbacks=callbacks)
```

Here I'm invoking EarlyStopping and setting it to monitor 'val_loss'. It will stop training when validation loss hits a minimum.

(taken from a simple demo here: https://github.com/BlankAdventure/snippets/blob/main/Python/autoenc_denoise.py)

**Reply by ecetsn** [reply]
Hello

Thank you for your answer, i was looking for an answer in terms of SKTime. I know that i can monitor val_loss with EarlyStopping from Keras but since the SKTime Neural Network Classifier's fit function does not take validation data as an input, I am little confused on how to do it. 


```
early_stopping = keras.callbacks.EarlyStopping( monitor='loss', patience=100, verbose=1, restore_best_weights=True)
reduce_lr = keras.callbacks.ReduceLROnPlateau(monitor='loss', factor=0.2, patience=50, min_lr=1e-6)

lstmfcn = LSTMFCNClassifier(n_epochs=2000, batch_size=128, dropout=0.8, 
                            kernel_sizes=(8, 5, 3), filter_sizes=(128, 256, 128), 
                            lstm_size=8, attention=False, callbacks=[early_stopping, reduce_lr], 
                            random_state=None, verbose=1)

lstmfcn.fit(X_train, y_train)  
```


This is how I am trying to deal with it at the moment but it would better if I could use validation data for the purpose. I am new at using SKTime and Keras. It would be great if I could see the learning curve.

Best

**Reply by BlankAdventure** [reply]
Maybe @fkiraly can comment, but, if some new features are needed for this I would be open to contributing.

**Reply by GabrielMSchmidt** [reply]
Hello, I'm having the same trouble as you. 

I want to train a InceptionTime Classifier with sktime but the .fit method does not accept validation data as an input, so if you found the answer, I would be happy to see it.
Thanks!
